New, better CleanerML docu

Tobias-B-Besemer opened the pull request:
Just to have a PR already open

Tobias-B-Besemer commented:
@az0 I forged you repo to my project https://github.com/LV-Crew as repo https://github.com/LV-Crew/docs and made you an admin...
Would be nice if we can work together on it!
I can't finish this alone!

Tobias-B-Besemer commented:
This is how far I am: https://github.com/LV-Crew/docs/blob/CleanerML-v2.0-Docu/_posts/cleanerml_v2.0.md

az0 commented:
I read over many parts. I agree that much useful information is not explicitly documented.

A general question is how this fits in with the [existing documentation about CleanerML](https://docs.bleachbit.org/doc/cleanerml.html)? The pull request does not replace it because the existing documentation has more information about environment variables, for example.

Another question is what is the best way to collaborate. I could either make suggestions for you to fix, but it would probably be faster for me to accept everything and then I change it. Sorry, I often like to make many changes. In this way, I also am a perfectionist.

A general observation is it could be helpful to start splitting up the long page into multiple pages. For example, there could be one for environment variables. (I have more to write about them that isn't yet documented.) Another page could about the `search` attribute, another about the `command` attribute, etc. My suggestion to move forward is to focus on just one of these pages to focus on. I could start environment variables, and you could start either `search` or `command`?

Let me know what you think. Again, thank you for your significant effort, Tobias.

Tobias-B-Besemer commented:
As I started with the docu, it was just some examples... Now I'm at the point where I think it would be best to merge the existing docu into the new one and replace the old one with the new one...

> Another question is what is the best way to collaborate...

Did you read my comment that I made you admin in the new repo I made? Did you had a view on the latest PR in this repo? You can just push your changes there and I get them! I edit my docu directly in GitHub Desktop with Notepad++, so I get informed about changes and N++ ask me to reload the page... ;-)

> A general observation is it could be helpful to start splitting up the long page into multiple pages...

Yeah, I thought this, too! And then I was thinking "Hey, why don't we use the GitHub Wiki ???"... ;-)

> Again, thank you for your significant effort, Tobias.

No prob, you're welcome! ;-)
Btw.: I like your project and IMHO it should be time to get more equal to CCleaner & Co. ... ;-)

Tobias-B-Besemer commented:
Oh, btw.: Best in collaboration would be (for me), I made some "Drafts" and you finish them. The reason therefor (at least) is, that my English is really not the best (at least for docus) and you should have a look if all is correct what I write and if I have something forgotten...

az0 commented:
You mentioned you made drafts and would like me to finish them, so how about you open a new pull request that creates a new article called "variables"? 

Topics include
* environment variables
* multi-value variables
* special variables

I'll look over it. Instead of asking you for changes, I'll probably accept it and make the edits myself

az0 commented:
More ideas for variables
* Variable syntax: $unix, %%windows%%, and $$multi$$
* Where to use them (not in .ini)
* On Windows, BleachBit defines %cd% ( current directory), commonappdata, documents, localappdata( windows xp), music, pictures, video, localappdatalow

Tobias-B-Besemer commented:
My plan was to work (and finish) on this docu as one of my next points...
My plan was glob, the variables and integrate the "old" docu into this one...
As I'm not sure about how I will structure all, my plan was to let the docu first as one page, restructure & integrate, and then split...
Just give me some time and you will see how things go...

Tobias-B-Besemer commented:
Andrew, today I'm worked some hours on the documentation...
(I wanna have it finished in the next 2-3 weeks, then the docu is also ready for the release of BB3.0... ^^)

This is what I have done so far:
- Restructured the document
- Formatted it better
- Added an "Content of the chapter"
- Added relative "jumping" links
- Added more information to "glob"
- Added the "Summary - Table with deletion use cases"
- Added the information from the VLC Cleaner (still raw)
- Added the information from your comments above (still raw)
- Added more information (still raw)

You can have a look at the current version here:
https://github.com/LV-Crew/bleachbit-docus/blob/CleanerML-v2.0-Docu/_posts/cleanerml_v2.0.md

Next I will do:
- Adding link to the open issue about wildcard in INI/Registry in the wishlist
- Adding link to the issue "Should have XML Cleaner"
- Adding link to the issue "command delete folder"

You can make already a first Review, make suggestions, add comments, or write something I can include.

By the work came the following question up:
This is code from `firefox.xml`...
`<running type="pathname">~/.mozilla/firefox/*.default/lock</running>`
Shouldn't be there a `search="glob"` ???
Or isn't `search="glob"` not supported in `running` ???
And if not, does it work ???

Tobias-B-Besemer commented:
`<action command="delete" search="deep" cache="false" regex="\.[Ll][Oo][Gg]$"/>`
What does in this example the option `cache="false"` do?
What would change if I set it to `cache="true"` ???

Tobias-B-Besemer commented:
And what does this means:
`regex="\.log(\.[0-9]+)?$`

File extension is `.log.[0-9]` + ???

Tobias-B-Besemer commented:
0-9 also maybe more or no digit at all ???

Tobias-B-Besemer commented:
And: Does recursive deletion also work with `regex=` and other `search=` like `walk.all`, `glob` & `file` ???

And when I can delete recursive with `regex=` and at least `search=walk.files` for what is (or where is the difference to) `search="deep"` ???

I should have answers to all the questions that I can document it...

az0 commented:
><running type="pathname">~/.mozilla/firefox/*.default/lock</running>
>Shouldn't be there a search="glob" ???
>Or isn't search="glob" not supported in running ???
>And if not, does it work ???
It's always glob

>What does in this example the option cache="false" do?
This feature was never fully implemented, so I removed it

>And what does this means:
>regex="\.log(\.[0-9]+)?$

It means the log ends in a number like `auth.log.1`. This is the standard on Linux for rotated logs

>And: Does recursive deletion also work with regex= and other search= like walk.all, glob & file ???

Yes, sure. However, it doesn't make sense to use it with `file` because it matches a single, literal file.

>And when I can delete recursive with regex= and at least search=walk.files for what is (or where is the difference to) search="deep" ???

`search="deep"` searches across the file system, currently just the user profile, and it is optimized for speed to search for multiple patterns at once.

`walk.files` matches all files (but not directories) under a path that is specified in the `path` attribute

Tobias-B-Besemer commented:
Thx! I will document that! :-)

Btw.:
> > ~/.mozilla/firefox/*.default/lock
> > Shouldn't be there a search="glob" ???
> > Or isn't search="glob" not supported in running ???
> > And if not, does it work ???
>
> It's always glob

Then this part of the code in the cleaner never worked... ^^

> > And what does this means:
> > regex=".log(.[0-9]+)?$
>
> It means the log ends in a number like `auth.log.1`. This is the standard on Linux for rotated logs

I'm not sure about the `+` and `?` at the end...

az0 commented:
>Then this part of the code in the cleaner never worked... ^^

I think it always worked. It does expand the glob.

>I'm not sure about the + and ? at the end...

The plus will match one or more such as `1` or `99` and the question mark makes the number optional

https://regex101.com/r/Vf2TIE/2

Tobias-B-Besemer commented:
Thank you very much for your answer! I will docu that...

Tobias-B-Besemer commented:
Docued! ^^
